src/libopensc/pkcs15-nqApplet.c

```diff
@@ -0,0 +1,238 @@
+	pin_info.attrs.pin.type = SC_PKCS15_PIN_TYPE_UTF8;
+	pin_info.attrs.pin.min_length = 6;
+	pin_info.attrs.pin.stored_length = 6;
+	pin_info.attrs.pin.max_length = 6;
+	pin_info.attrs.pin.pad_char = '\0';
+	pin_info.tries_left = -1; //TODO
+	pin_info.max_tries = 3;
+
+	strlcpy(pin_obj.label, "UserPIN", sizeof(pin_obj.label));
+	pin_obj.flags = SC_PKCS15_CO_FLAG_MODIFIABLE | SC_PKCS15_CO_FLAG_PRIVATE;
+
+	rv = sc_pkcs15emu_add_pin_obj(p15card, &pin_obj, &pin_info);
+	LOG_TEST_RET(card->ctx, rv, "sc_pkcs15emu_add_pin_obj failed");
+
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+static int add_nqapplet_certificate(sc_pkcs15_card_t *p15card, const char *id, const char *name, u8 data_id)
+{
+	int rv;
+	struct sc_pkcs15_cert_info cert_info;
+	struct sc_pkcs15_object cert_obj;
+	sc_card_t *card = p15card->card;
+
+	LOG_FUNC_CALLED(card->ctx);
+
+	memset(&cert_info, 0, sizeof(cert_info));
+	memset(&cert_obj, 0, sizeof(cert_obj));
+
+	sc_pkcs15_format_id(id, &cert_info.id);
+	cert_info.authority = 0;
+	rv = get_nqapplet_certificate(card, data_id, &cert_info.value);
+	LOG_TEST_RET(card->ctx, rv, "Failed to get certificate");
+
+	strlcpy(cert_obj.label, name, sizeof(cert_obj.label));
+	cert_obj.flags = 0;
+
+	rv = sc_pkcs15emu_add_x509_cert(p15card, &cert_obj, &cert_info);
+	LOG_TEST_RET(card->ctx, rv, "sc_pkcs15emu_add_x509_cert failed");
+
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+static int add_nqapplet_private_key(sc_pkcs15_card_t *p15card, const char *id, int reference, const char *name, const char *pin_id, unsigned int usage)
+{
+	int rv;
+	struct sc_pkcs15_prkey_info prkey_info;
+	struct sc_pkcs15_object prkey_obj;
+	sc_card_t *card = p15card->card;
+
+	LOG_FUNC_CALLED(card->ctx);
+
+	memset(&prkey_info, 0, sizeof(prkey_info));
+	memset(&prkey_obj, 0, sizeof(prkey_obj));
+
+	sc_pkcs15_format_id(id, &prkey_info.id);
+	prkey_info.usage = usage;
+	prkey_info.native = 1;
+	prkey_info.key_reference = reference;
+	prkey_info.modulus_length = 3072;
+
+	strlcpy(prkey_obj.label, name, sizeof(prkey_obj.label));
+	prkey_obj.flags = SC_PKCS15_CO_FLAG_PRIVATE;
+	sc_pkcs15_format_id(pin_id, &prkey_obj.auth_id);
+
+	rv = sc_pkcs15emu_add_rsa_prkey(p15card, &prkey_obj, &prkey_info);
+	LOG_TEST_RET(card->ctx, rv, "sc_pkcs15emu_add_rsa_prkey failed");
+
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+static int add_nqapplet_objects(sc_pkcs15_card_t *p15card)
+{
+	int rv;
+	sc_card_t *card = p15card->card;
+
+	LOG_FUNC_CALLED(card->ctx);
+
+	// 1) User PIN
+	rv = add_nqapplet_pin(p15card, "1", 0x01);
+	LOG_TEST_RET(card->ctx, rv, "Failed to add PIN 1");
+
+	// 2.1) CH.Auth
+	rv = add_nqapplet_certificate(p15card, "1", "C.CH.Auth", 0x00);
+	LOG_TEST_RET(card->ctx, rv, "Failed to add Auth. certificate");
+
+	// 2.2) PrK.CH.Auth
+	rv = add_nqapplet_private_key(p15card, "1", 0x01, "PrK.CH.Auth", "1", SC_PKCS15_PRKEY_USAGE_SIGN | SC_PKCS15_PRKEY_USAGE_ENCRYPT | SC_PKCS15_PRKEY_USAGE_DECRYPT);
```

Comment:
I think the `SC_PKCS15_PRKEY_USAGE_ENCRYPT` should not be here as its not a private key operation.

Reply:
You are right, removed it.